### PR TITLE
Fix rFactor when there are indirect root domain(s), and refactor

### LINKED ISSUE #1692: Inconsistent rfactor
In this fusion, the rfactor tensor gets an inconsistent rfactor domain.

```
TEST_F(NVFuserTest, TMP) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  auto tv0 = makeSymbolicTensor(2);
  fusion.addInput(tv0);

  auto tv1 = sum(tv0, {0, 1});

  fusion.addOutput(tv1);

  tv1->split(1, 4);
  tv1->split(0, 3);
  tv1->merge(1, 2);
  auto rf = tv1->rFactor({-1});

  std::cout << ir_utils::toString(rf->getRootDomain()) << std::endl;
  std::cout << ir_utils::toString(rf->getMaybeRFactorDomain()) << std::endl;

  GpuLower gpu_lower(&fusion);
}
```

The rfactor domain of tensor `rf`: `iS9{i1} iS15{( 3 * ( ceilDiv(i2, 4) ) )}rf rS14{4}rf`. Specifically, `3` comes from `i1`, but it's also included in the second ID.

Not sure if that's the root cause, but the fusion results in an error when building the computeAt map:

```
* thread #1, name = 'test_jit', stop reason = breakpoint 1.1
    frame #0: 0x00007fffd5fd910b libstdc++.so.6`__cxxabiv1::__cxa_throw(obj=0x0000000002d9c720, tinfo=0x00007fffd60927a8, dest=(libstdc++.so.6`std::out_of_range::~out_of_range() at stdexcept.cc:65:33))(void *)) at eh_throw.cc:77:1
Process 13953 launched: '/home/nmaruyama/pytorch/debug3/build/bin/test_jit' (x86_64)
(lldb) bt
* thread #1, name = 'test_jit', stop reason = breakpoint 1.1
  * frame #0: 0x00007fffd5fd910b libstdc++.so.6`__cxxabiv1::__cxa_throw(obj=0x0000000002d9c720, tinfo=0x00007fffd60927a8, dest=(libstdc++.so.6`std::out_of_range::~out_of_range() at stdexcept.cc:65:33))(void *)) at eh_throw.cc:77:1
    frame #1: 0x00007fffd5fd6087 libstdc++.so.6`std::__throw_out_of_range(__s="_Map_base::at") at functexcept.cc:82:5
    frame #2: 0x00007ffff6f5c9ce libtorch_cuda.so`std::__detail::_Map_base<torch::jit::fuser::cuda::IterDomain*, std::pair<torch::jit::fuser::cuda::IterDomain* const, torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain*, std::hash<torch::jit::fuser::cuda::IterDomain*> > >, std::allocator<std::pair<torch::jit::fuser::cuda::IterDomain* const, torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain*, std::hash<torch::jit::fuser::cuda::IterDomain*> > > >, std::__detail::_Select1st, std::equal_to<torch::jit::fuser::cuda::IterDomain*>, std::hash<torch::jit::fuser::cuda::IterDomain*>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<false, false, true>, true>::at(this=0x0000000002d98418, __k=0x00007fffffffbe38) at hashtable_policy.h:760:2
    frame #3: 0x00007ffff6f55fdd libtorch_cuda.so`std::unordered_map<torch::jit::fuser::cuda::IterDomain*, torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain*, std::hash<torch::jit::fuser::cuda::IterDomain*> >, std::hash<torch::jit::fuser::cuda::IterDomain*>, std::equal_to<torch::jit::fuser::cuda::IterDomain*>, std::allocator<std::pair<torch::jit::fuser::cuda::IterDomain* const, torch::jit::fuser::cuda::VectorOfUniqueEntries<torch::jit::fuser::cuda::IterDomain*, std::hash<torch::jit::fuser::cuda::IterDomain*> > > > >::at(this=0x0000000002d98418, __k=0x00007fffffffbe38) at unordered_map.h:991:21
    frame #4: 0x00007ffff6f52a3b libtorch_cuda.so`torch::jit::fuser::cuda::IterDomainGraph::build(this=0x0000000002d982f0, fusion=0x0000000002d28140) at compute_at_map.cpp:180:22
```

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes https://github.com/csarofeen/pytorch/issues/1692

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 1]
```diff
@@ -19824,6 +19824,40 @@ TEST_F(NVFuserTest, FusionRfactorPredication2_CUDA) {
...
+  auto tv0 = makeSymbolicTensor(3);
+  fusion.addInput(tv0);
+
+  auto tv1 = sum(tv0, {1, 2});
+  fusion.addOutput(tv1);
+
+  tv1->split(2, 4);
+  tv1->split(1, 3);
+  tv1->merge(2, 3);
+  auto rf = tv1->rFactor({-1});
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:19841]:
```C++
fusion.print()
```
```
%kernel {
T2_l[ iS11{i1}, iS14{( ceilDiv(i2, 3) )}, iS18{( 3 * ( ceilDiv(i3, 4) ) )}rf, rS17{4}rf ]
   = reduction( T0_g[ iS0{i1}, iS1{i2}, iS2{i3} ], op = add, initial value = double(0), allreduce = 0 )
T1_g[ iS19{i1}, rS22{( ceilDiv(i2, 3) )}, rS24{( 3 * ( ceilDiv(i3, 4) ) )} ]
   = reduction( T2_l[ iS11{i1}, iS14{( ceilDiv(i2, 3) )}, iS18{( 3 * ( ceilDiv(i3, 4) ) )}rf, rS17{4}rf ], op = add, initial value = double(0), allreduce = 0 )

TransformPrinter : 
T0_g[ iS0{i1}, iS1{i2}, iS2{i3} ]
 root domain : (iS0{i1},iS1{i2},iS2{i3})
T2_l[ iS11{i1}, iS14{( ceilDiv(i2, 3) )}, iS18{( 3 * ( ceilDiv(i3, 4) ) )}rf, rS17{4}rf ]
 root domain : (iS11{i1},iS12{i2},rS13{i3}rf)
  Split: rS13{i3}rf by factor 4 -> iS16{( ceilDiv(i3, 4) )}rf, rS17{4}rf, start offset: 0, stop offset: 0
 rfactor domain : (iS11{i1},iS12{i2},iS16{( ceilDiv(i3, 4) )}rf,rS17{4}rf)
  Split: iS12{i2} by factor 3 -> iS14{( ceilDiv(i2, 3) )}, iS15{3}, start offset: 0, stop offset: 0
  Merge: iS15{3} and iS16{( ceilDiv(i3, 4) )}rf -> iS18{( 3 * ( ceilDiv(i3, 4) ) )}rf
T1_g[ iS19{i1}, rS22{( ceilDiv(i2, 3) )}, rS24{( 3 * ( ceilDiv(i3, 4) ) )} ]
 root domain : (iS19{i1},rS20{i2},rS21{( ceilDiv(i3, 4) )})
  Split: rS20{i2} by factor 3 -> rS22{( ceilDiv(i2, 3) )}, rS23{3}, start offset: 0, stop offset: 0
  Merge: rS23{3} and rS21{( ceilDiv(i3, 4) )} -> rS24{( 3 * ( ceilDiv(i3, 4) ) )}
}
```

**Comment by csarofeen** [comment]
@zasdfgbnm I pushed a more significant refactor of transform rfactor, please review and let me know what you think.

**Review by csarofeen — APPROVED** [review]
Approving, but feel free to continue refactoring/reviewing @zasdfgbnm

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**